### PR TITLE
Sentry gun UI

### PR DESCRIPTION
**Pull request by mullenpaul**
# About the pull request

This PR takes the idea of the sentry laptop from the aliens film and introduces the ability to configure defensive structures. The bulk of the work here is to introduce a framework to allow other contributors to build weird and wonderful ways to interact with turrets in new ways.

The sentry laptop is a control point which can be used by trained engineers to link sentry guns and other defensive structures for monitoring and configuration.

The laptop as well as several radio keys are available in req.

## effects on sentry guns

- sentry guns can have their IFF status configured, allowing for USCM, USCM+WY and all humanoid (everything non xeno)
- some sentry guns can have their rate of fire changed, at the effect of accuracy: single fire, burst fire and full-auto
- sentry guns can act like cameras (not connected to the camera network) and have a 'gunsight' camera, allowing an operator to see what is in their kill zone

## effects on multitool

The multitool is used to link the sentry laptop to a sentry gun. The multitool gets a reference to the laptop and when used on the sentry gun it links the two. A sentry gun can only be linked to one laptop. When a sentry gun is linked to a laptop it is considered 'encrypted' and can't be dismantled until unlinked - this encourages linked sentry guns to be used in a static role.

## effects on radio

This PR introduces a new radio channel which has its own encryption key. When the sentry gun is engaged it will announce to the radio channel (CD 60 seconds).

## setup instructions

- spawn sentry laptop
- place on table
- use multitool on laptop to load encryption keys
- deploy sentry gun
- use multitool on sentry gun
- access laptop

## screenshots
On startup
image

Configuration page
image

Status page
image

Overview Page
image

Sentry 'camera gunsight'
image

## Looking for feedback

- How do we distribute these laptops? We can have req have a few available
- How common do we want the encryption keys to be? There is a focus on relying on others for communication so they probably shouldn't be too common.

# Explain why it's good for the game

Allows for more interaction with sentry guns from the engineer. Builds a framework to allow for interesting interactions with the sentry without tgui changes.

# Testing Photographs and Procedure

<details>
<summary>Screenshots & Videos</summary>

Put screenshots and videos here with an empty line between the screenshots and the `<details>` tags.

</details>

# Changelog

:cl:
add: added sentry laptop to game
add: new radio channel for sentry gun updates, only listenable on radio headsets
balance: added method to configure rate of fire of sentry guns at cost of accuracy
refactor: allowed for the configuration of IFF groups of sentry guns
refactor: multitools now can store encryption keys
refactor: added UI to manage sentry guns
imageadd: added sentry laptop sprite
/:cl:

### CODE REVIEW

**Comment by NoKohi** [comment]
shouldn't it be USCM and not USMC
same thing with multitools, they got renamed to security access tuners

File: code/game/objects/items/devices/radio/encryptionkey.dm [hunk 1]
```diff
@@ -35,6 +35,12 @@
 	icon_state = "cap_key"
 	channels = list(RADIO_CHANNEL_ALMAYER = 1, RADIO_CHANNEL_COMMAND = 1, RADIO_CHANNEL_MP = 1, RADIO_CHANNEL_ENGI = 1, RADIO_CHANNEL_MEDSCI = 1, RADIO_CHANNEL_REQ = 1, SQUAD_MARINE_1 = 1, SQUAD_MARINE_2 = 1, SQUAD_MARINE_3 = 1, SQUAD_MARINE_4 = 1, SQUAD_MARINE_5 = 1, SQUAD_MARINE_CRYO = 0, RADIO_CHANNEL_JTAC = 1, RADIO_CHANNEL_INTEL = 1)
 
+/obj/item/device/encryptionkey/sentry_laptop
+	name = "Sentry Network Status Encryption Key"
+	desc = "Automated channel to broadcast sentry gun updates"
+	icon_state = "eng_key"
+	channels = list(RADIO_CHANNEL_SENTY = 1)
```
Comment (stanalbatross) [inline code/game/objects/items/devices/radio/encryptionkey.dm:42]:
```suggestion
	channels = list(RADIO_CHANNEL_SENTY = TRUE)
```

File: code/modules/defenses/defenses.dm [hunk 2]
```diff
@@ -148,6 +159,33 @@
 			to_chat(user, SPAN_WARNING("You don't have the training to do this."))
 			return
 
+		if(friendly_faction(user.faction))
+			var/obj/item/device/multitool/tool = O
+			if(length(tool.encryption_keys) > 0)
+				if(linked_laptop == null)
```
Comment (stanalbatross) [inline code/modules/defenses/defenses.dm:165]:
```suggestion
				if(!linked_laptop)
```

File: code/modules/defenses/sentry.dm [hunk 3]
```diff
@@ -22,10 +23,29 @@
...
 	handheld_type = /obj/item/defenses/handheld/sentry
 
+	var/enaged_timer = null
+	var/low_ammo_timer = null
+	var/sent_empty_ammo = FALSE
+
+	// action list is configurable for all subtypes, this is just an example
+	choice_categories = list(
+		"RATE OF FIRE" = list("SINGLE", "BURST", "FULL-AUTO"),
+		"IFF STATUS" = list("USCM", "WY", "HUMAN"),
```
Comment (stanalbatross) [inline code/modules/defenses/sentry.dm:41]: use defines for "single" etc instead of strings

File: code/modules/defenses/sentry_computer.dm [hunk 4]
```diff
@@ -0,0 +1,420 @@
...
+	var/atom/movable/screen/map_view/cam_screen
+	var/atom/movable/screen/background/cam_background
+
+	/// The turf where the camera was last updated.
+	var/turf/last_camera_turf
+
+	/// All turfs within range of the currently active camera
+	var/list/range_turfs = list()
+
+/datum/camera_holder/proc/Setup(var/mapName)
```
Comment (stanalbatross) [inline code/modules/defenses/sentry_computer.dm:15]: snake_case for DM...

File: code/modules/defenses/sentry_computer.dm [hunk 5]
```diff
@@ -0,0 +1,420 @@
...
+	cam_background.assigned_map = map_name
+	cam_background.del_on_map_removal = FALSE
+
+/datum/camera_holder/Destroy(force, ...)
+	. = ..(force)
+	qdel(cam_background)
+	qdel(cam_screen)
+
+/obj/item/device/sentry_computer
+	name = "Sentry Computer"
```
Comment (stanalbatross) [inline code/modules/defenses/sentry_computer.dm:32]:
```suggestion
	name = "sentry computer"
```

File: code/modules/defenses/sentry_computer.dm [hunk 6]
```diff
@@ -0,0 +1,420 @@
...
+		to_chat(user, SPAN_NOTICE("The new cell contains: [new_cell.charge] power."))
+		cell.forceMove(get_turf(user))
+		cell = new_cell
+		user.drop_inv_item_to_loc(new_cell, src)
+		playsound(src,'sound/machines/click.ogg', 25, 1)
+	else if(istype(object, /obj/item/device/multitool))
+		var/obj/item/device/multitool/tool = object
+		var/id = tool.serial_number
+		playsound(src, get_sfx("terminal_type"), 25, FALSE)
+		if (do_after(usr, 2, INTERRUPT_NO_NEEDHAND, BUSY_ICON_GENERIC))
```
Comment (stanalbatross) [inline code/modules/defenses/sentry_computer.dm:195]: 0.2 SECONDS

any value below 0.6 seconds does not trigger a do_after I believe (need to check this)

File: code/modules/defenses/sentry_computer.dm [hunk 7]
```diff
@@ -0,0 +1,420 @@
...
+				registered_tools -= list(id)
+			else
+				to_chat(user, SPAN_NOTICE("You load an encryption key to the [tool]."))
+				registered_tools += list(id)
+				tool.load_encryption_key(serial_number, src)
+	else
+		..()
+
+/obj/item/device/sentry_computer/proc/register(var/tool, mob/user, var/sentry_gun)
+	if (do_after(user, 2, INTERRUPT_NO_NEEDHAND, BUSY_ICON_GENERIC))
```
Comment (stanalbatross) [inline code/modules/defenses/sentry_computer.dm:207]: same

File: code/modules/defenses/sentry_computer.dm [hunk 8]
```diff
@@ -0,0 +1,420 @@
...
+		var/obj/structure/machinery/defenses/sentry/sentry = sentry_gun
+		pair_sentry(sentry)
+		to_chat(user, SPAN_NOTICE("\The [sentry] has been encrypted."))
+		var/message = "[sentry] added to [src]"
+		INVOKE_ASYNC(src, .proc/send_message, message)
+
+/obj/item/device/sentry_computer/proc/unregister(var/tool, mob/user, var/sentry_gun)
+	var/obj/structure/machinery/defenses/sentry/sentry = sentry_gun
+	if(sentry.linked_laptop == src)
+		if (do_after(user, 2, INTERRUPT_NO_NEEDHAND, BUSY_ICON_GENERIC))
```
Comment (stanalbatross) [inline code/modules/defenses/sentry_computer.dm:217]: same

File: code/modules/defenses/sentry_computer.dm [hunk 9]
```diff
@@ -0,0 +1,420 @@
...
+	.["mapRef"] = map_name
+	var/index = 1
+	for(var/sentry in paired_sentry)
+		var/list/sentry_holder = list()
+		var/obj/structure/machinery/defenses/sentry/sentrygun = sentry
+		sentry_holder["selection_menu"] = list()
+		sentry_holder["index"] = index
+		sentry_holder["name"] = sentrygun.name
+		sentry_holder["health_max"] = sentrygun.health_max
+		index += 1
```
Comment (stanalbatross) [inline code/modules/defenses/sentry_computer.dm:271]:
```suggestion
		index++
```

File: code/modules/defenses/sentry_computer.dm [hunk 10]
```diff
@@ -0,0 +1,420 @@
...
+	.["camera_target"] = null
+
+	var/index = 1
+
+	for(var/sentry in paired_sentry)
+		var/list/sentry_holder = list()
+		var/obj/structure/machinery/defenses/defence = sentry
+		if(current == defence)
+			.["camera_target"] = index
+		index += 1
```
Comment (stanalbatross) [inline code/modules/defenses/sentry_computer.dm:295]:
```suggestion
		index++
```

File: code/modules/defenses/sentry_flamer.dm [hunk 11]
```diff
@@ -9,9 +9,29 @@
...
 
+	choice_categories = list(
+		"RATE OF FIRE" = list("SINGLE", "FULL-AUTO"),
+		"IFF STATUS" = list("USMC", "WY", "HUMAN"),
+	)
+
+	selected_categories = list(
+		"RATE OF FIRE" = "SINGLE",
+		"IFF STATUS" = "USMC",
+	)
```
Comment (stanalbatross) [inline code/modules/defenses/sentry_flamer.dm:20]: defines...

**Review by stanalbatross — CHANGES_REQUESTED** [review]
all I caught in initial code pass

**Review by stanalbatross — CHANGES_REQUESTED** [review]
image
error on turning on sentry while window was still open
[MAY DISREGARD IF THIS IS ALREADY FIXED, I THINK I'M ON AN OLD VERSION]

File: code/modules/defenses/sentry_computer.dm (hunk 6 above)
Comment (mullenpaul) [reply code/modules/defenses/sentry_computer.dm:195]: I've set all these to 10, I didn't realise this was in deciseconds

File: code/modules/defenses/sentry.dm (hunk 3 above)
Comment (mullenpaul) [reply code/modules/defenses/sentry.dm]: I've moved all of these strings to defines

**Comment by mullenpaul** [comment]
I've resolved the first round of code reviews

File: code/modules/defenses/defenses.dm [hunk 12]
```diff
@@ -73,6 +82,8 @@
 	message += "\n"
 	if(display_additional_stats)
 		message += SPAN_INFO("Its display reads - Kills: [kills] | Shots: [shots].")
+	if(length(nickname) > 0)
+		message += SPAN_INFO("It has been assigned the name: [nickname]")
```
Comment (stanalbatross) [inline code/modules/defenses/defenses.dm:86]:
```suggestion
		message += SPAN_INFO("\nIt has been assigned the name: [nickname]")
```

**Review by stanalbatross — CHANGES_REQUESTED** [review]
image
when linking to a plasma sentry
image
weird artefact?
image
this font is also too small
image
you can scroll the tabs up and down with the scroll wheel for some reason
image
not every sentry's base ammo is the same - this should take into account % of mag size
image
strange empty space here?
image
you can link tesla coils to it which causes this runtime when trying to swap iff types

you can also link mounted flags and change their iff... which makes not much sense

**Comment by mullenpaul** [comment]
I made a mistake on a rebase and ended up deleting the branch. I have recreated the PR here https://github.com/cmss13-devs/cmss13/pull/2100

**Pull request closed without merging**